pyccl/pk2d.py

```diff
@@ -109,26 +120,56 @@ def __init__(self, pkfunc=None, a_arr=None, lk_arr=None, pk_arr=None,
             if (len(a_arr)*len(lk_arr) != len(pkflat)):
                 raise ValueError("Size of input arrays is inconsistent")
         else:  # Initialize power spectrum from function
-            # Check that the input function has the right signature
-            try:
-                pkfunc(k=np.array([1E-2, 2E-2]), a=0.5)
-            except Exception:
-                raise ValueError("Can't use input function")
-
+            warnings.warn("The use of a function when initialising a ``Pk2D`` "
+                          "object is deprecated. Use `Pk2D.from_function`.",
+                          CCLDeprecationWarning)
             # Set k and a sampling from CCL parameters
             a_arr = get_pk_spline_a(cosmo=cosmo)
             lk_arr = get_pk_spline_lk(cosmo=cosmo)
 
             # Compute power spectrum on 2D grid
-            pkflat = np.array([pkfunc(k=np.exp(lk_arr), a=a) for a in a_arr])
-            pkflat = pkflat.flatten()
+            pkflat = np.array([pkfunc(k=np.exp(lk_arr), a=a)
+                               for a in a_arr]).flatten()
 
         status = 0
         self.psp, status = lib.set_pk2d_new_from_arrays(lk_arr, a_arr, pkflat,
                                                         int(extrap_order_lok),
                                                         int(extrap_order_hik),
                                                         int(is_logp), status)
-        check(status, cosmo=cosmo)
+        check(status)
+
+    @classmethod
+    def from_function(cls, pkfunc, *, is_logp=True, cosmo=None,
+                      extrap_order_lok=1, extrap_order_hik=2):
+        """ Generates a `Pk2D` object from a function that calculates a power
+        spectrum.
+
+        Args:
+            pkfunc : :obj:`callable`
+                Function with signature ``f(k, a)`` which takes vectorized
+                input in ``k`` (wavenumber in :math:`\\mathrm{Mpc}^{-1}``)
+                and a scale factor`a``, and returns the value of the
+                corresponding quantity. ``pkfunc`` will be sampled at the
+                values of ``k`` and ``a`` used internally by CCL to store the
+                linear and non-linear power spectra.
+            cosmo : :class:`~pyccl.core.Cosmology`, optional cosmological
+                parameters. Used to determine sampling rates in scale factor
+                and wavenumber.
+
+        Returns:
+            :class:`~pyccl.pk2d.Pk2D`
+                Power spectrum object.
+        """
+        # Set k and a sampling from CCL parameters
+        a_arr = get_pk_spline_a(cosmo=cosmo)
+        lk_arr = get_pk_spline_lk(cosmo=cosmo)
+
+        # Compute power spectrum on 2D grid
+        pk_arr = np.array([pkfunc(k=np.exp(lk_arr), a=a) for a in a_arr])
```

Comment:
And since we are creating this all-shiny new function, why don't we force users to vectorize it outside of CCL?
Best-case scenario, we don't have to run this slow loop. Worst-case scenario, users have to add an extra line of code before calling this function.

Reply:
I'd rather not force users to do this, for the simple reason that we haven't done so so far, and this will only make the transition even more painful than it's already going to be. This can be done later on (check if vectorised, otherwise we vectorise).

Reply:
That's a fair point, although you were about to completely deprecate `pkfunc`.

FYI, doing it in the future won't work without breaking API though. The way to check if a function is vectorized is either by adding an extra argument that users have to declare very time, or by try/except.
In the try/except block, if the function isn't vectorized and `len(k_arr) == len(a_arr)`, `numpy` will automatically broadcast them together, and do a pairwise calculation, which is really **not** what the user intended for.

So, if we ever want to implement this, and save us from running Python loops, the best time is now. Otherwise it's going to have to stay like this until v4.

Reply:
Since we're not even sure this functionality is very useful, I'd rather minimize pain for those who use it now.

And we are completely deprecating pkfunc!